### PR TITLE
Go: reinstate models-as-data sink conversions with fixes

### PR DESCRIPTION
**Pull request by owen-mc**
The first 14 commits are reinstating commits that were reverted in https://github.com/github/codeql/pull/17296. Then there are some commits fixing things: reverting some models back to QL and adding some models-as-data models for `logrus.FieldLogger`. Then there are some commits adding tests that would have caught the problems in the first place. Finally, there are some commits adding a heuristic for logger calls to replace results that we now miss because we have converted all logging models to MaD (because QL models normally use `Method.getACall()`, which is too broad and matches any interface method which the modeled method implements).

### CODE REVIEW

File: csharp/ql/campaigns/Solorigate/lib/change-notes/2024-11-20-heuristic-logging-sinks.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+category: minorAnalysis
+---
+* A call to a method whose name starts with "Debug", "Error", "Fatal", "Info", "Log", "Output", "Panic", "Print", "Trace", "Warn" or "With" defined on an interface whose name ends in "logger" or "Logger" is now considered a LoggerCall. In particular, it is a sink for `go/clear-text-logging` and `go/log-injection`. This may lead to some more alerts in those queries.
```
Comment (michaelnebel) [inline csharp/ql/campaigns/Solorigate/lib/change-notes/2024-11-20-heuristic-logging-sinks.md:4]: Is this the right location for the change note?

**Comment by owen-mc** [comment]
I've now looked through the QA results from ~5,000 repos. We get a lot of extra results for log injection (~1,500) and cleartext logging (~300). I sampled them and they all seem to be valid results from us adding a heuristic for local logger interfaces. I looked in detail at all the repos where we lost results. (We lost  ~40 results in total.) Some were because they are calling logger functions using a variable, which isn't currently supported. I shouldn't be too hard but there may be a performance penalty. I will file a follow-up issue to look into that. I also found a bug in my recent work to fix models-as-data inheritance, which I will fix as a follow-up. There are also a handful of lost alerts because we were previously matching something we hadn't actually modeled because of the known issue where `Function.getACall()` (which is routinely used in QL models) also matches calls to an interface method which the function implements. This means that we are accidentally matching some libraries that we haven't modeled, just because they have similarities with libraries that we have modelled. 

Overall I think these results are very good, and the handful of lost alerts shouldn't stop this PR from being merged.

**Comment by smowton** [comment]
Change note needs moving as @michaelnebel notes; then happy to merge per that description.

File: csharp/ql/campaigns/Solorigate/lib/change-notes/2024-11-20-heuristic-logging-sinks.md (hunk 1 above)
Comment (owen-mc) [reply csharp/ql/campaigns/Solorigate/lib/change-notes/2024-11-20-heuristic-logging-sinks.md]: For some reason VS code is always trying to recommend this directory for change notes, and I finally fell for it 🤦🏻 .

**Pull request merged**